server/src/main/java/org/folio/reservoir/server/IngestWriteStream.java

```diff
@@ -71,17 +71,15 @@ public Future<Void> write(JsonObject globalRecord) {
         }
         return future;
       })
-      .onComplete(x -> {
-        if (ops.decrementAndGet() == queueSize / 2 && drainHandler != null) {
-          drainHandler.handle(null);
-        }
-        if (ops.get() == 0) {
-          log.info("{} records processed", number.get());
-          if (endHandler != null) {
+        .onComplete(x -> {
+          if (ops.decrementAndGet() == queueSize / 2 && drainHandler != null) {
+            drainHandler.handle(null);
+          }
+          if (ops.get() == 0 && endHandler != null) {
+            log.info("{} records processed", number.get());
```

Comment:
Does this work when endHandler is never assigned?

Reply:
Correct. endHandler served two purposes. Fixed by introducing an ended boolean.